### ISSUE TITLE
Not wrapping templates with double quotes should fail gracefully

### ISSUE THREAD

**Issue by cmeury** [roboll/helmfile#92]
When an environment variable template in a value field is not wrapped in double quotes:

```yaml
set:
  - name: vault.replicas 
    value: {{ env "NUM_REPLICAS" }}
```
`helmfile` will just use an empty string:

    exec: helm upgrade --install [..] --set vault.replicas=

As per the discussion in #88, we should probably warn the user about this; to prevent a frustrating error hunt.

It looks like different YAML parsers handle this differently. Apparently the golang one tolerates the double braces. The following two do not validate the above snippet:

* https://jsonformatter.org/yaml-formatter
* https://codebeautify.org/yaml-validator

**Comment by sstarcher** [comment]
Would it make sense to just interpolate the entire helmfile instead of interpolating specific items as we currently do?

This would also allow for people to wrap segments of the file as conditionals.

I can see a benefit of doing it both ways.

**Comment by mumoshu** [comment]
@sstarcher I'm still considering if that's a good thing to do. YAML in general doesn't get along well with templating mainly because of indentation issue like seen in many `{{ foo | indent }}` across many helm charts...

**Comment by mumoshu** [comment]
Can't we just limit the templating capability of helmfile to that of today, and just enable helmfile to consume any helmfile.yaml'ish input via standard input like `yourtemplateengine helmfile.yml.tmpl | helmfile sync -f -` so that you can leverage any template engine-ish things? (btw jsonnet is my favorite 😉

**Comment by mumoshu** [comment]
In that case, I'm open to accept any contributions to a "best practices" doc or examples to encourage users to leverage helmfile as a composable tool used within your deployment process.

My sentiment is that we should support as many use-cases as possible while not bloating up helmfile's code-base. IMHO, templating is especially what people vary in opinions.

**Comment by sstarcher** [comment]
Templating out the entire file instead of sections would likely simplify the code base along with adding additional capabilities.  Helm itself templates out entire files so they don't seem to think that `{{ foo | indent }}` is that large of an issue.

I would overall lean toward templating the entire file.

One benefit of templating only sections would be if it had extra template functionality like allowing the chart name, version, etc(can't think of anything else) to be injected as template variables, but I believe templating the entire file would add an even greater capability.

**Comment by mumoshu** [comment]
> Templating out the entire file instead of sections would likely simplify the code base along with adding additional capabilities.

True!

> Helm itself templates out entire files so they don't seem to think that {{ foo | indent }} is that large of an issue.

Maybe. It is just that I have seen several people in helm and kube-aws community including me(!) complained about the yaml+templating issue. But I'm sure not everyone has the same sentiment.

> I would overall lean toward templating the entire file.

I understand the necessity to solve real problems. It is that just I can't help wishing that we could find a better solution 😢 

> One benefit of templating only sections would be if it had extra template functionality like allowing the chart name, version, etc(can't think of anything else) to be injected as template variables

Good point. That could be something that we can never implement when we fully moved to templating the entire file. In that regard, it wouldn't be the final solution if we moved to the entire file templating and omitted current templating capability?

**Comment by sstarcher** [comment]
An example that could be solved by only individual templating instead of global templating would be as follows.

Currently I have lots of hardcoded names that would be great if I could simply the file paths, but I'm guessing we don't want to support any specific magic folder structure in helmfile.

```
  - name: chart1
    namespace: ops
    chart: stable/chart1
    version: 0.2.0
    values:
      - helm/releases/chart1/values.yaml
    secrets:
      - helm/account/{{ env "HELM_ACCOUNT" }}/chart1/secrets.yaml
  - name: chartOther
    namespace: ops
    chart: stable/chartOther
    version: 0.1.0
    values:
      - helm/releases/chartOther/values.yaml
      - helm/account/{{ env "HELM_ACCOUNT" }}/chartOther/values.yaml
    secrets:
      - helm/account/{{ env "HELM_ACCOUNT" }}/chartOther/secrets.yaml
```

Individual templating could improve this somewhat by 
```
  - name: chart1
    namespace: ops
    chart: stable/chart1
    version: 0.2.0
    values:
      - helm/releases/{{ .Name }}/values.yaml
    secrets:
      - helm/account/{{ env "HELM_ACCOUNT" }}/{{ .Name }}/secrets.yaml
  - name: chartOther
    namespace: ops
    chart: stable/chartOther
    version: 0.1.0
    values:
      - helm/releases/{{ .Name }}/values.yaml
      - helm/account/{{ env "HELM_ACCOUNT" }}/{{ .Name }}/values.yaml
    secrets:
      - helm/account/{{ env "HELM_ACCOUNT" }}/{{ .Name }}/secrets.yaml
```

Also looking for any other recommendations on how I could solve the above from being repeated 20 times in my helmfile as it increased the length and complexity considerably.

**Comment by sstarcher** [comment]
ya, we could certainly support both if we desired via a mechanism like `{{}}` is global and say `[[]]`. 

I'm also up for solving it an entirely different way if we want to deviate from how helm does things if we can make a good case for a better solution.

**Comment by cmeury** [comment]
Also, it's Helm people running this tool, they already know about the challenges of YAML+go-template ;)

**Comment by mumoshu** [comment]
Thanks for all the feedbacks! I may have been too nervous to go ahead with yaml+templates but yeah, as we are already helm users, it shouldn't be a huge problem anyway.

Note that helm v3 is going to support altenative template engines so we may also end up in following the same path in the future, though. But I'm now convinced that making helmfile universal to any template engine as I've suggested would be a kind of premature optimizations at this point in time and in this context!

So, I'm open to review any PRs towards the direction. Let's implement the feature that templating the entire file. Please beware not to break backward-compatibility though, as we already have prod users with a bit different use-case.

Formely helmfile had been almost a k8s manifest for a set of releases. It will evolve to something like a helm template of releases after this improvement :)

**Comment by mumoshu** [comment]
Closing this as resolved via #98
